plugins/modules/cloud/misc/proxmox_instance_actions.py

```diff
@@ -0,0 +1,313 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# Copyright: (c) 2020, Tristan Le Guern <tleguern at bouledef.eu>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: proxmox_instance_actions
+short_description: Perform actions on Qemu/KVM and LXC instances from Proxmox VE
+version_added: 2.0.0
+description:
+  - Perform actions on an existing instance from Proxmox VE.
+  - This module does not return data.
+options:
+  action:
+    description:
+      - Perform the given action.
+      - C(reset) is only valid with Qemu/KVM instances.
+    type: str
+    choices: ['reboot', 'reset', 'resume', 'shutdown', 'start', 'stop', 'suspend']
```

Comment:
Hmm, this *looks* quite different from how Ansible modules usually operate, namely be specifying a state and making sure the object in question satisfies that state.

It might be better to name the module `proxmox_instance` with a `state` parameter which has values like `rebooted`, `resetted`, `shutdowned`, `started`, `stopped`, `suspended`. This module already compares the actions to the current state, so essentially the behavior that users would expect is already there, it's mainly the way the interface works right now that looks very differently.

CC @Thulium-Drake @resmo for your opinions on this :)

Reply:
I'm a bit split, I understand the desire to use a uniform style for all modules. But which one to choose, that's a difficult one and mainly a discussion of preference and semantics.

So on the one hand we have the Openstack modules that use this style and on the other hand we have the 'object' style (used by lxc_container and the vmware modules).

I do lean more towards going 'object' style, as that is the general way to manage 'things' with Ansible, be it services or virtual machines.

Reply:
The AWS modules and the docker modules also use states and not actions.

Also os_server_action seems to be the only module following that convention (at least the only one I found); a newer OS one, baremetal_node_action, uses `state` instead (https://opendev.org/openstack/ansible-collections-openstack/src/branch/master/plugins/modules/baremetal_node_action.py).

Reply:
Fair enough, let's stick with states then :-)

With regards to the name of the module, I think ```proxmox_instance``` is a good name :-)

Reply:
I choose actions for this module because they do not map one-to-one with states (as shown in the matrice in comments) and allow subtler behaviours.

For me `resetted`, `rebooted` and `restarted` are not states, only mere descriptions of past events. The real state is most likely `running`, aka `present` in most Ansible modules.
Also using states rather than actions it is not possible to distinguish between `shutdowned` and `stopped` because the end result is the same: the vm is off, which is a state.

Reply:
@Aversiste indeed, they are not proper states. A lot of modules either use such pseudo-states (see for example the `service` module), or have boolean flags like `restart` to trigger restarts (for example `docker_container`).

Reply:
Perhaps look at it this way:

- reboot = restarted
- reset = restarted + force
- shutdown = shutdown
- poweroff = shutdown + force
- suspend = suspended
- resume = running

And is it a problem that we can't differentiate between shutdown and poweroff?

Reply:
@Thulium-Drake the difference between `shutdown` and `stop` is important: the former is graceful and the later is forceful ;)
And in situations where a VM is stuck only `stop` can help.

If you both think it is too alien compared to the rest of Ansible modules then I won't push further. In my opinion it doesn't make sense to convert this module to the state model as it is already implemented in both `proxmox_kvm` and `proxmox`, I propose to retire this pull request.

Reply:
@Aversiste are all actions here already covered by proxmox_kvm? If not, adding the missing ones there would definitely be a good idea.

About this module: it feels a bit like the OS actions module was a kind of 'mistake', since I haven't seen that pattern repeated anywhere yet.

What might be of interest though is a action plugin like `reboot` for proxmox instances (which also waits for the machine to come back). It could offer to reset, restart, or reboot depending on a flag/choice option.

Reply:
@felixfontein not all actions are covered by proxmox_kvm, no. Regarding the action plugin I do not know how to write one and did not found documentation about it. Do you have an example I could use?

By the way issue #1906 just opened touching on the current subject.

Am I understanding correctly then that I should rather rework the handling of the `state` parameter in proxmox_kvm? I am thinking about adding `state=paused` using the suspend API call, `state=stopped` + `force=no` using shutdown, `state=stopped` + `force=yes` using stop, `state=restarted` + `force=no` using reboot, `state=restarted` + `force=yes` using reset and finally `state=running` using resume or start depending on the situation.

Reply:
Proxmox_kvm also have another limitations, like you can't clone and assign a MAC, ie. you just can't modify the net device, and neither does it seem to be able to add disks, (My use case is to have a base template that gets cloned, CPU/RAM parameters set, assigned a specific MAC, network, tag (for DHCP statics) and then the needed storage added.
So the question is rather: should it be split/refactor the various functions out of proxmox_kvm (and proxmox(lxc)) into run_actions, config_changes, clone_actions etc. or would the beterer be to enhance/extend the proxmox_kvm functionality as above.
As it presently stands, I'll have to hack some scripts/URI-api calls together to get the results I need (and more tech debt) or help spent some effort to extend/refactor to get those functionality required

Reply:
@hvisage splitting proxmox_kvm in smaller modules like I tried here is not the way to go. In my proxmox role I do similar tricks although I ssh into the node and run various qm commands rather than using the API. Improving `update` is definitely an old task in my todo list.

Reply:
Hmm, I think splitting up the functionality, or at least having dedicated modules for some of the functionality would be a great step. It's just that the interface of this specific implementation is something that doesn't conform to Ansible's semantics. Having for example a module `proxmox_instance_state` which just does state handling would be OK. Whether it's worth the effort of maintaining both `proxmox_kvm` and the same state changes in `proxmox_instance_state` is another question (which eventually has to be answered by the maintainers).

Reply:
@felixfontein in this case I prefer something in module_utils shared between `proxmox_kvm` and `proxmox` :)

Reply:
@Aversiste yes, that's definitely a good idea ;)